src/thor/timedep_forward.cc

```diff
@@ -0,0 +1,285 @@
+       !costing_->Allowed(directededge, pred, tile, edgeid, localtime) ||
+        costing_->Restricted(directededge, pred, edgelabels_, tile,
+                                     edgeid, true, localtime)) {
+      continue;
+    }
+
+    // Skip shortcut edges when near the destination. Always skip within
+    // 10km but also reject long shortcut edges outside this distance.
+    // TODO - configure this distance based on density?
+    if (directededge->is_shortcut() && (pred.distance() < 10000.0f ||
+        directededge->length() > max_shortcut_length)) {
+      continue;
+    }
+
+    // Update the_shortcuts mask
+    shortcuts |= directededge->shortcut();
+
+    // Compute the cost to the end of this edge
+    Cost newcost = pred.cost() + costing_->EdgeCost(directededge) +
+          costing_->TransitionCost(directededge, nodeinfo, pred);
+
+    // If this edge is a destination, subtract the partial/remainder cost
+    // (cost from the dest. location to the end of the edge).
+    auto p = destinations_.find(edgeid);
+    if (p != destinations_.end()) {
+      // Subtract partial cost and time
+      newcost -= p->second;
+
+      // Find the destination edge and update cost to include the edge score.
+      // Note - with high edge scores the convergence test fails some routes
+      // so reduce the edge score.
+      for (const auto& destination_edge : destination.path_edges()) {
+        if (destination_edge.graph_id() == edgeid) {
+          newcost.cost += destination_edge.distance();
+        }
+      }
+      newcost.cost = std::max(0.0f, newcost.cost);
+
+      // Mark this as the best connection if that applies. This allows
+      // a path to be formed even if the convergence test fails (can
+      // happen with large edge scores)
+      if (best_path.first == -1 || newcost.cost < best_path.second) {
+         best_path.first = (es->set() == EdgeSet::kTemporary) ?
+            es->index() : edgelabels_.size();
+        best_path.second = newcost.cost;
+      }
+    }
+
+    // Check if edge is temporarily labeled and this path has less cost. If
+    // less cost the predecessor is updated and the sort cost is decremented
+    // by the difference in real cost (A* heuristic doesn't change)
+    if (es->set() == EdgeSet::kTemporary) {
+      EdgeLabel& lab = edgelabels_[es->index()];
+      if (newcost.cost <  lab.cost().cost) {
+        float newsortcost = lab.sortcost() - (lab.cost().cost - newcost.cost);
+        adjacencylist_->decrease(es->index(), newsortcost);
+        lab.Update(pred_idx, newcost, newsortcost);
+      }
+      continue;
+    }
+
+    // If this is a destination edge the A* heuristic is 0. Otherwise the
+    // sort cost (with A* heuristic) is found using the lat,lng at the
+    // end node of the directed edge.
+    float dist = 0.0f;
+    float sortcost = newcost.cost;
+    if (p == destinations_.end()) {
+      const GraphTile* t2 = directededge->leaves_tile() ?
```

Comment:
Unrelated performance thought - is there enough room in the node ID to add a enough additional information that fetching a tile could be avoided here, and the heuristic would still be admissible?  Avoiding a tile fetch here could make for a big speedup.

Reply:
would need to store the lat,lon of the end node for each edge so the A* heuristic can be computed. Since a small percentage of the edges leave the tile I think the small cost for occasional lookups offsets the larger impact to tile size.

Reply:
By "admissible", I meant in the technical sense - https://en.wikipedia.org/wiki/Admissible_heuristic .  Something along the lines of `if the node is not on this tile, use the closest point on the target tile`.  Would be interesting to measure whether the tradeoff of a less efficient search (due to the poorer heuristic) would outperform avoiding potentially expensive tile fetching.

Reply:
I've been burned enough through the years on A* heuristics not being quite right that I don't think it is worth it. For one, the majority case is that a tile does not need to be fetched. Even if it is, there is a high probability that it will be needed in the future when/if expanding from the end node of this edge - so the memory is likely to be required later anyway. It is hard to predict how this impacts memory cache behavior but I am hesitant to grab spare bits from the directed edge.